worker/worker.go

```diff
@@ -1174,6 +1174,147 @@ func (w *worker) objectsHandlerPUT(jc jape.Context) {
 	}
 }
 
+func (w *worker) objectsHandlerPackedPUT(jc jape.Context) {
+	jc.Custom((*[]byte)(nil), nil)
+	ctx := jc.Request.Context()
+
+	// fetch the request.
+	req := []api.PackedObjectInfo{}
+	dec := json.NewDecoder(jc.Request.Body)
+	if jc.Check("failed to decode object infos", dec.Decode(&req)) != nil {
+		return
+	}
+	requestBody := io.MultiReader(dec.Buffered(), jc.Request.Body)
+
+	// fetch the upload parameters
+	up, err := w.bus.UploadParams(ctx)
+	if jc.Check("couldn't fetch upload parameters from bus", err) != nil {
+		return
+	}
+
+	// cancel the upload if consensus is not synced
+	if !up.ConsensusState.Synced {
+		w.logger.Errorf("upload cancelled, err: %v", api.ErrConsensusNotSynced)
+		jc.Error(api.ErrConsensusNotSynced, http.StatusServiceUnavailable)
+		return
+	}
+
+	// allow overriding the redundancy settings
+	rs := up.RedundancySettings
+	if jc.DecodeForm(queryStringParamMinShards, &rs.MinShards) != nil {
+		return
+	}
+	if jc.DecodeForm(queryStringParamTotalShards, &rs.TotalShards) != nil {
+		return
+	}
+	if jc.Check("invalid redundancy settings", rs.Validate()) != nil {
+		return
+	}
+
+	// allow overriding contract set
+	var contractset string
+	if jc.DecodeForm(queryStringParamContractSet, &contractset) != nil {
+		return
+	} else if contractset != "" {
+		up.ContractSet = contractset
+	}
+
+	// attach gouging checker to the context
+	ctx = WithGougingChecker(ctx, up.GougingParams)
+
+	// attach contract spending recorder to the context.
+	ctx = WithContractSpendingRecorder(ctx, w.contractSpendingRecorder)
+
+	w.pool.setCurrentHeight(up.CurrentHeight)
+	usedContracts := make(map[types.PublicKey]types.FileContractID)
+
+	// fetch contracts
+	contracts, err := w.bus.Contracts(ctx)
+	if jc.Check("couldn't fetch contracts from bus", err) != nil {
+		return
+	}
+
+	// randomize order of contracts so we don't always upload to the same hosts
+	frand.Shuffle(len(contracts), func(i, j int) { contracts[i], contracts[j] = contracts[j], contracts[i] })
+
+	// keep track of slow hosts so we can avoid them in consecutive slab uploads
+	slow := make(map[types.PublicKey]int)
+
+	// generate a unique key for each object.
+	keys := make([]object.EncryptionKey, len(req))
+	var crs []io.Reader
+	for i, r := range req {
+		keys[i] = object.GenerateEncryptionKey()
+		crs = append(crs, keys[i].Encrypt(io.LimitReader(requestBody, int64(r.Length))))
+	}
+	cr := io.MultiReader(crs...)
+
+	var objectSize int
+	var slabs []object.Slab
+	for {
```

Comment:
if we are adding this handler it feels like we should probably dedupe all of the now duplicated upload logic into a separate method so changes to the upload code will be reflected in both the packing and regular upload route